packages/ember-data/lib/system/references/has-many.js

```diff
@@ -0,0 +1,84 @@
+import Ember from 'ember';
+import Reference from './reference';
+
+const get = Ember.get;
+
+var HasManyReference = function(store, parentInternalModel, hasManyRelationship) {
+  this._super$constructor(store, parentInternalModel);
+  this.hasManyRelationship = hasManyRelationship;
+  this.type = hasManyRelationship.relationshipMeta.type;
+
+  if (hasManyRelationship.link) {
+    this.remoteType = "link";
+    this.link = hasManyRelationship.link;
+  } else {
+    this.remoteType = "ids";
+    this._ids = hasManyRelationship.canonicalState.map(function(internalModel) {
+      return internalModel.id;
+    });
+  }
+
+  // TODO parent
+  // TODO inverse
+};
+
+HasManyReference.prototype = Ember.create(Reference.prototype);
+HasManyReference.prototype.constructor = HasManyReference;
+HasManyReference.prototype._super$constructor = Reference;
+
+HasManyReference.prototype.ids = function() {
+  return this._ids;
+};
+
+HasManyReference.prototype.meta = function() {
+  return this.hasManyRelationship.manyArray.meta;
+};
+
+HasManyReference.prototype.push = function(objectOrPromise) {
+  return Ember.RSVP.resolve(objectOrPromise).then((data) => {
+    var internalModels = data.map((payload) => {
```

Comment:
Unsure whether we should support arrays here, probably, but we should at least support json api payload format where it is in `data:`